flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/proxy/DynamoDbStreamsProxy.java

```diff
@@ -0,0 +1,184 @@
+public class DynamoDbStreamsProxy implements StreamProxy {
+
+    private final DynamoDbStreamsClient dynamoDbStreamsClient;
+    private final SdkHttpClient httpClient;
+    private final Map<String, String> shardIdToIteratorStore;
+
+    private static final FullJitterBackoff BACKOFF = new FullJitterBackoff();
+
+    private static final Logger LOG = LoggerFactory.getLogger(DynamoDbStreamsProxy.class);
+
+    public DynamoDbStreamsProxy(DynamoDbStreamsClient dynamoDbStreamsClient, SdkHttpClient httpClient) {
+        this.dynamoDbStreamsClient = dynamoDbStreamsClient;
+        this.httpClient = httpClient;
+        this.shardIdToIteratorStore = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public List<Shard> listShards(String streamArn, @Nullable String lastSeenShardId) {
+        return this.getShardsOfStream(streamArn, lastSeenShardId);
+    }
+
+    @Override
+    public GetRecordsResponse getRecords(String streamArn, String shardId, StartingPosition startingPosition) {
+        String shardIterator =
+                shardIdToIteratorStore.computeIfAbsent(
+                        shardId, (s) -> getShardIterator(streamArn, s, startingPosition));
+
+        if (shardIterator == null) {
+            return null;
+        }
+        try {
+            GetRecordsResponse getRecordsResponse = getRecords(shardIterator);
+            if (getRecordsResponse.nextShardIterator() != null) {
+                shardIdToIteratorStore.put(shardId, getRecordsResponse.nextShardIterator());
+            }
+            return getRecordsResponse;
+        } catch (ExpiredIteratorException e) {
+            // Eagerly retry getRecords() if the iterator is expired
+            shardIterator = getShardIterator(streamArn, shardId, startingPosition);
+            GetRecordsResponse getRecordsResponse = getRecords(shardIterator);
+            if (getRecordsResponse.nextShardIterator() != null) {
+                shardIdToIteratorStore.put(shardId, getRecordsResponse.nextShardIterator());
+            }
+            return getRecordsResponse;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        dynamoDbStreamsClient.close();
+        httpClient.close();
+    }
+
+    private List<Shard> getShardsOfStream(
+            String streamName, @Nullable String lastSeenShardId) {
+        List<Shard> shardsOfStream = new ArrayList<>();
+
+        DescribeStreamResponse describeStreamResponse;
+        do {
+            describeStreamResponse = this.describeStream(streamName, lastSeenShardId);
+            List<Shard> shards = describeStreamResponse.streamDescription().shards();
+            for (Shard shard : shards) {
+                shardsOfStream.add(shard);
+            }
+
+            if (shards.size() != 0) {
+                lastSeenShardId = shards.get(shards.size() - 1).shardId();
+            }
+        } while (!CollectionUtils.isEmpty(describeStreamResponse.streamDescription().shards()));
```

Comment:
We can use `LastEvaluatedShardId` to evaluate exit condition of loop too

Reply:
yeah, i think i picked it up from existing DDB streams adapter code. Will fix this